File: set_matrix_zeroes.js
var setZeroes = function (matrix) {
  let col0 = 1;
  for (let row = 0; row < matrix.length; row++) {
    if (matrix[row][0] === 0) { col0 = 0; }
    for (let col = 1; col < matrix[row].length; col++) {
      const num = matrix[row][col];
      
      if(num === 0) {
        matrix[0][col] = matrix[row][0] = 0;
      }
    } 
  }

  for (let row = matrix.length - 1; row >= 0; row--) {
    for (let col = matrix[row].length - 1; col > 0; col--) {
      if(matrix[row][0] === 0 || matrix[0][col] === 0) {
        matrix[row][col] = 0;
      }
    }
    if (col0 === 0 ) {matrix[row][0] = 0;}
  }


  return matrix;
};

console.log(setZeroes([
  [1, 1, 1],
  [0, 1, 1],
  [1, 1, 1]
]));

File: Set Matrix Zeroes .swift
//
//  Set Matrix Zeroes .swift
//  
//
//  Created by Nix on 2/12/19.
//

import Foundation

class Solution {
    func setZeroes(_ matrix: inout [[Int]]) {
        var zeroInFirstRow: Bool = false
        var zeroInFirstCol: Bool = false
        
        for row in 0..<matrix.count {
            for col in 0..<matrix[0].count {
                if matrix[row][col] == 0 {
                    if row == 0 { zeroInFirstRow = true }
                    if col == 0 { zeroInFirstCol = true }
                    matrix[0][col] = 0
                    matrix[row][0] = 0
                }
            }
        }
        
        for col in 1..<matrix[0].count {
            if matrix[0][col] == 0 {
                for row in 1..<matrix.count {
                    matrix[row][col] = 0
                }
            }
            if zeroInFirstRow { matrix[0][col] = 0 }
        }
        
        for row in 1..<matrix.count {
            if matrix[row][0] == 0 {
                for col in 1..<matrix[0].count {
                    matrix[row][col] = 0
                }
            }
            if zeroInFirstCol { matrix[row][0] = 0 }
        }
    }
}
